### PR TITLE
Update type hints for all internal environments

### PR DESCRIPTION
**Pull request by robertoschiavone**
# Description

This commit adds strict type checking to all environments, comforming them to the documentation.

It shouldn't break anything and no new tests should be required.

## Type of change
# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Thanks for doing this, could you add `from __future__ import annotations` as the first line of code then run pre-commit. This will update the code to nicer type annotations in cases

**Review by pseudo-rnd-thoughts — APPROVED** [review]

File: gymnasium/envs/mujoco/mujoco_env.py [hunk 1]
```diff
@@ -309,7 +312,7 @@ def render(self):
 
     def _get_viewer(
         self, mode
-    ) -> Union["mujoco_py.MjViewer", "mujoco_py.MjRenderContextOffscreen"]:
+    ) -> mujoco_py.MjViewer | mujoco_py.MjRenderContextOffscreen:
```
Comment (Kallinteris-Andreas) [inline gymnasium/envs/mujoco/mujoco_env.py:315]: This requires `mujoco_py` to be imported,
if `mujoco_py` is not imported does this cause issues?

File: gymnasium/envs/mujoco/mujoco_env.py [hunk 2]
```diff
@@ -380,7 +383,7 @@ def __init__(
 
     def _initialize_simulation(
         self,
-    ) -> Tuple["mujoco._structs.MjModel", "mujoco._structs.MjData"]:
+    ) -> tuple[mujoco._structs.MjModel, mujoco._structs.MjData]:
```
Comment (Kallinteris-Andreas) [inline gymnasium/envs/mujoco/mujoco_env.py:386]: same here

File: gymnasium/envs/mujoco/mujoco_env.py (hunk 1 above)
Comment (Kallinteris-Andreas) [inline gymnasium/envs/mujoco/mujoco_env.py:315]: This requires `mujoco_py` to be imported,
if `mujoco_py` is not imported does this cause issues?

**Review by Kallinteris-Andreas — COMMENTED** [review]
besides those i am good with the `mujoco` envs

**Comment by pseudo-rnd-thoughts** [comment]
@robertoschiavone Could you test if @Kallinteris-Andreas comments do cause an issue with mujoco or mujoco-py is not installed

File: gymnasium/envs/mujoco/mujoco_env.py (hunk 1 above)
Comment (robertoschiavone) [reply gymnasium/envs/mujoco/mujoco_env.py:315]: No more than usual. There's an "import guard" (not sure it's the right way to call it) at the beginning of the file

```python
try:
    import mujoco_py
except ImportError as e:
    MUJOCO_PY_IMPORT_ERROR = e
else:
    MUJOCO_PY_IMPORT_ERROR = None
  ```

and later on an exception gets thrown

```python
        if MUJOCO_PY_IMPORT_ERROR is not None:
            raise error.DependencyNotInstalled(
                f"{MUJOCO_PY_IMPORT_ERROR}. "
                "(HINT: you need to install mujoco-py, and also perform the setup instructions "
                "here: https://github.com/openai/mujoco-py.)"
            )
```

Therefore `_get_viewer` is effectively unreachable if `mujoco` is not installed.

**Comment by RedTachyon** [comment]
I left a few comments inside of the code, but I'm gonna pull that out to the top level here.

I'm not 100% sure, but maybe like 90% sure that adding the type hints of `ObsType` etc. is actually harmful and breaks the type system.

Attaching two screenshots. One is PyCharm's type analysis with this PR merged:
image

The second one is after removing the `ObsType` type hint from `CartPoleEnv.reset`:
image


The conclusion is that with this merged, in `obs, info = env.reset()`, `obs` is only detected as an `ObsType` (whereas we know it should be `ndarray`), which is a generic without any bounds and not attached to anything, so it's pretty much a synonym for `Any`. We see this because the type analysis detects it as just `ObsType`, and doesn't complain about calling a non-existent `ndarray.test()` method.

Removing (part) of the changes of this PR results in correct behavior - `obs` is detected as `ndarray`, and trying to do `ndarray.test()` correctly raises a type warning.


@pseudo-rnd-thoughts I'd strongly consider reverting this and maybe redoing it in a smaller, more careful scope. We don't need to have explicit type hints for every subclass, that's what the generic is for, but if we want to add them, we have to add them correctly.

File: gymnasium/envs/mujoco/mujoco_env.py (hunk 1 above)
Comment (RedTachyon) [reply gymnasium/envs/mujoco/mujoco_env.py:315]: I think this is only true due to `from __future__ import annotations`, so it's not necessarily a problem in this case, but we need to be careful about it.

**Comment by pseudo-rnd-thoughts** [comment]
@RedTachyon I think your are right partially but we just need replace `ObsType` and `ActType` with `np.ndarray` in most cases plus `Env[np.ndarray, np.ndarray]` in the `class` definition

**Comment by RedTachyon** [comment]
Yea, my point was that adding explicit `np.ndarray` type hints (or the equivalent for some envs) is probably harmless, but fundamentally inconsequential for the type system if `Env[np.ndarray, np.ndarray]` is also specified. But the way it's done now, it doesn't work (also `SupportsFloat` shouldn't be used here, since these specific envs just return a float)

**Comment by pseudo-rnd-thoughts** [comment]
@RedTachyon I'm happy to add a REWARD type with float (maybe int and np.ndarray) for v1.0.0

**Comment by RedTachyon** [comment]
@pseudo-rnd-thoughts I think these are two separate issues? Even if the base annotation on `Env` is some `Reward`, however it would be defined, the specific environments can/should still be annotated as simply `float`, since that's literally what they return, we don't need any indirection

**Pull request merged**